### PR TITLE
feat(llmobs): agent attribution on spans

### PR DESCRIPTION
**Pull request by yahya-mouman**
## Description

Ports LLMObs **agent attribution** to the Go tracer, on par with [dd-trace-py#18788](https://github.com/DataDog/dd-trace-py/pull/18788) and [dd-trace-js#9175](https://github.com/DataDog/dd-trace-js/pull/9175).

Every LLMObs span that has an agent ancestor now carries:

```json
"meta": {
  "agent_attribution": {
    "pagent_name":    "<nearest agent ancestor name>",
    "pagent_span_id": "<that agent span's id>"
  }
}
```

Spans with no agent ancestor omit the block entirely.

## What's implemented

- **Resolve + store** (`internal/llmobs/llmobs.go`): nearest agent resolved once at `StartSpan` via a one-level lookup that inherits the parent's already-resolved attribution (O(1), no ancestor walk). An agent never attributes itself; a sub-agent attributes to its enclosing agent.
- **Serialize** (`internal/llmobs/llmobs.go`): surfaced as `meta.agent_attribution` at span finish, gated on presence. `pagent_name` is JSON `null` when id-only, matching the Python/Node wire shape.
- **Wire safety** (`internal/llmobs/context.go`): `AgentNameWireSafe` rejects bytes outside `[0x20, 0x7E]`, comma, semicolon, and tilde. Semicolons and tildes are excluded for W3C tracecontext compatibility (`composeTracestate` sanitizes them to `_`, which would corrupt the name downstream). Length is not gated here; the budget check happens at the call site.
- **Distributed propagation** (`ddtrace/tracer/span.go`, `ddtrace/tracer/propagating_tags.go`): wired into the existing `setLLMObsPropagatingTags` function alongside the other LLMObs propagating tags. Writes `_dd.p.llmobs_pagent_span_id` (always when there's an agent ancestor) and `_dd.p.llmobs_pagent_name` (only when wire-safe and the current tag set has budget left). Both keys are explicitly unset when there's no agent ancestor so stale values from a previous sibling can't leak downstream. Budget is checked against `DefaultMaxTagsHeaderLen` using a `propagatingTagsByteLen()` helper that matches the measurement `marshalPropagatingTags` uses internally.

## Ownership boundary

Changes are in `internal/llmobs/` and `ddtrace/tracer/`. No changes to the `APMSpan` interface or any non-LLMObs/tracer package.

## Testing

- **In-process:** tool under agent, indirect nesting (`agent → workflow → tool`), sub-agent under agent (nearest wins), top-level agent/LLM omit block, no-agent-in-chain omits block
- **Distributed** (in `TestStartSpan/distributed-context-propagation-*`, all via `testClientServer`): agent upstream propagates attribution, attribution forwarded through non-agent intermediary, no-agent omits, stale attribution cleared when non-agent sibling makes the outbound call, wire key names (`_dd.p.llmobs_pagent_span_id` / `_dd.p.llmobs_pagent_name`), id-only when name is wire-unsafe
- Full suite green: `go test ./internal/llmobs/... ./ddtrace/tracer/...`, `go vet` clean

## Follow-up (minor, non-blocking)

- Public `InjectContext`/`ExtractContext` API for cases where automatic APM propagation isn't available (deferred to a follow-up PR)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/llmobs/start_span_agent_test.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (rarguelloF) [inline internal/llmobs/start_span_agent_test.go]: there's no need to create separate test files for these new tests, they can go into the existing ones based on where the tested functionality file name exists.

File: internal/llmobs/resolve_parent_agent_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (rarguelloF) [inline internal/llmobs/resolve_parent_agent_test.go]: same here

File: internal/llmobs/util.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (rarguelloF) [inline internal/llmobs/util.go]: nit: same here, this can go in the same file where its used.

File: internal/llmobs/propagation.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	span, ok := ActiveLLMSpanFromContext(ctx)
+	if !ok || span == nil {
+		return
+	}
+	injectAgentAttribution(span, carrier)
+}
+
+// ExtractContext reads agent attribution from the carrier and attaches it to a
+// PropagatedLLMSpan on the returned context, preserving any fields already set.
+func ExtractContext(ctx context.Context, carrier map[string]string) context.Context {
```
Comment (rarguelloF) [inline internal/llmobs/propagation.go]: I see these 2 functions, despite having generic names (Inject/Extract Context), don't do anything with the base LLMObs spanID/traceID propagation info. Is this intended?

Also, base propagation works out of the box by passing contexts around into the `StartSpan` functions. Can't we do the same for these new fields? It should be a simple change/addition over the existing mechanisms.

Finally, if we conclude this manual inject/extract API is actually necessary, let's do it in a separate PR, make it consistent with the core tracer APIs, have it cover the base span info (trace/span/ml_app), not just the new attribution fields, and have extensive documentation/examples on when and how to use.

**Review by rarguelloF — CHANGES_REQUESTED** [review]

File: internal/llmobs/propagation.go (hunk 4 above)
Comment (yahya-mouman) [reply internal/llmobs/propagation.go]: Fair point on all of this. Removing InjectContext/ExtractContext from this PR entirely.

The in-process part (resolveParentAgent at StartSpan + meta.agent_attribution serialization) is self-contained and that's what I'll ship here. The distributed case deserves its own PR anyway.

File: internal/llmobs/util.go (hunk 3 above)
Comment (yahya-mouman) [reply internal/llmobs/util.go]: Done, deleted along with propagation.go.

File: internal/llmobs/resolve_parent_agent_test.go (hunk 2 above)
Comment (yahya-mouman) [reply internal/llmobs/resolve_parent_agent_test.go]: Done, merged into start_span_agent_test.go.

File: internal/llmobs/start_span_agent_test.go (hunk 1 above)
Comment (yahya-mouman) [reply internal/llmobs/start_span_agent_test.go]: Done, consolidated both test files into one.

File: internal/llmobs/propagation.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	parts = append(parts, tagKeyLLMObsPAgentSpanID+"="+spanID)
+
+	// Append the name only if wire-safe and within the total length budget.
+	if name != "" && agentNameWireSafe(name) {
+		candidate := append(parts, tagKeyLLMObsPAgentName+"="+name)
+		if joined := strings.Join(candidate, ","); len(joined) <= maxDatadogTagsLen {
+			parts = candidate
+		}
+	}
+	carrier[headerKeyDatadogTags] = strings.Join(parts, ",")
```
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply internal/llmobs/propagation.go]: All valid bugs, but they're all in InjectContext/ExtractContext which I'm removing from this PR (see the thread above).

File: internal/llmobs/propagation.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// extractAgentAttribution parses the carrier's x-datadog-tags value for the
+// llmobs agent attribution tags. present is true if the span id tag was found.
+func extractAgentAttribution(carrier map[string]string) (name string, spanID string, present bool) {
+	header, ok := carrier[headerKeyDatadogTags]
```
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply internal/llmobs/propagation.go]: All valid bugs, but they're all in InjectContext/ExtractContext which I'm removing from this PR (see the thread above).

[the remark this answers is not in the document]
Comment (yahya-mouman) [reply internal/llmobs/propagation.go]: All valid bugs, but they're all in InjectContext/ExtractContext which I'm removing from this PR (see the thread above).

File: llmobs/llmobs.go [hunk 7]
```diff
@@ -451,3 +451,22 @@ func parseAnnotateOptions(opts ...AnnotateOption) illmobs.SpanAnnotations {
 func isSpanKind(s Span, target illmobs.SpanKind) bool {
 	return illmobs.SpanKind(s.Kind()) == target
 }
+
+// InjectContext writes the active LLMObs span's agent attribution into the
+// carrier so a downstream service can resolve agent_attribution for its spans.
+// Call this alongside the APM tracer's Inject when making a distributed request.
```
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply llmobs/llmobs.go]: All valid bugs, but they're all in InjectContext/ExtractContext which I'm removing from this PR (see the thread above).

File: internal/llmobs/propagation.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	if spanID == "" {
+		return // no agent ancestor: write nothing
+	}
+
+	existing := carrier[headerKeyDatadogTags]
+	parts := make([]string, 0, 2)
+	if existing != "" {
+		parts = append(parts, existing)
+	}
+	parts = append(parts, tagKeyLLMObsPAgentSpanID+"="+spanID)
```
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply internal/llmobs/propagation.go]: All valid bugs, but they're all in InjectContext/ExtractContext which I'm removing from this PR (see the thread above).

File: internal/llmobs/propagation.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	// Update, do not overwrite: preserve any existing MLApp/TraceID/SpanID.
+	var prop PropagatedLLMSpan
+	if existing, ok := PropagatedLLMSpanFromContext(ctx); ok && existing != nil {
+		prop = *existing
+	}
+	prop.ParentAgentName = name
+	prop.ParentAgentSpanID = spanID
+	return ContextWithPropagatedLLMSpan(ctx, &prop)
```
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply internal/llmobs/propagation.go]: All valid bugs, but they're all in InjectContext/ExtractContext which I'm removing from this PR (see the thread above).

File: internal/llmobs/start_span_agent_test.go (hunk 1 above)
Comment (rarguelloF) [reply internal/llmobs/start_span_agent_test.go]: unresolving as this test file still exists.

File: internal/llmobs/start_span_agent_test.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	spanID  string
+	traceID string
+}
+
+func (s testAPMSpan) Finish(_ FinishAPMSpanConfig)          {}
+func (s testAPMSpan) AddLink(_ SpanLink)                    {}
+func (s testAPMSpan) SpanID() string                        { return s.spanID }
+func (s testAPMSpan) TraceID() string                       { return s.traceID }
+func (s testAPMSpan) SetBaggageItem(_ string, _ string)     {}
+func (s testAPMSpan) BaggageItem(_ string) string           { return "" }
```
Comment (rarguelloF) [inline internal/llmobs/start_span_agent_test.go]: I don't think there's a need to test using mock spans. Prefer to not use mocks unless its absolutely necessary.

**Review by rarguelloF — CHANGES_REQUESTED** [review]
With the newest changes there's no cross-process propagation of these new fields. Please check how things are currently being automatically propagated, it should be a simple addition in those places.

File: internal/llmobs/start_span_agent_test.go (hunk 1 above)
Comment (yahya-mouman) [reply internal/llmobs/start_span_agent_test.go]: deleted it! tests are now in `llmobs_test.go` going through the real tracer, asserting on `meta.agent_attribution` from the collector output. no more mock spans.

File: internal/llmobs/start_span_agent_test.go (hunk 10 above)
Comment (yahya-mouman) [reply internal/llmobs/start_span_agent_test.go]: agreed, removed them. all the same cases (tool-under-agent, indirect nesting, sub-agent, propagated parent, no-agent) are now in `llmobs_test.go` asserting on `meta.agent_attribution` from the collector.

**Comment by yahya-mouman** [comment]
I think I addresses all open issues !

**Comment by rarguelloF** [comment]
@codex review

File: ddtrace/tracer/span.go [hunk 11]
```diff
@@ -1347,6 +1347,12 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
 	} else {
 		spanCtx.trace.unsetPropagatingTag(keyPropagatedLLMObsSessionID)
 	}
+	if pagentID := llmSpan.PropagatedParentAgentSpanID(); pagentID != "" {
+		spanCtx.trace.setPropagatingTag(keyPropagatedLLMObsPAgentSpanID, pagentID)
+		if name := llmSpan.PropagatedParentAgentName(); illmobs.AgentNameWireSafe(name) {
+			spanCtx.trace.setPropagatingTag(keyPropagatedLLMObsPAgentName, name)
```
[the remark this answers is not in the document]
Comment (rarguelloF) [reply ddtrace/tracer/span.go]: +1, reproduced both halves on this branch. A root workflow with an agent sibling that publishes through an APM child, then a second workflow sibling making the outbound call, and the downstream tool span came back with `pagent_name: leaky_agent` despite having no agent in its chain. For the desync, an agent named `bad,name` following one named `safe_agent` sent the second agent's id paired with the first one's name.

There's a third variant for the same fix: an id-only inbound attribution writes an empty name, since `AgentNameWireSafe("")` returns true, and `isValidPropagatableTag` then rejects it at inject time and stamps `_dd.propagation_error=encoding_error` on the local root span. Unsetting the name whenever it isn't written, and unsetting both keys when there's no id, covers all three.

[the remark this answers is not in the document]
Comment (rarguelloF) [reply ddtrace/tracer/span.go]: +1, with numbers from this branch. A 256 byte name, which `AgentNameWireSafe` accepts, leaves the header at 493 bytes with no session ID, so it survives on 19 bytes of headroom. Adding a 40 character session ID drops it entirely, and the downstream span then reported `parent_id: undefined` with a fresh LLMObs trace ID, having also lost `_dd.p.dm` and `_dd.p.tid`.

dd-trace-py guards this in `_stamp_agent_attribution` with a 485 byte budget and a degrade path. Worth noting `propagatingTagsLen()` returns a count rather than a byte length, so the check needs to sum `len(k)+len(v)` over `iteratePropagatingTags`.

File: internal/llmobs/util.go [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+	if len(name) > 256 {
+		return false
+	}
+	for i := 0; i < len(name); i++ {
+		b := name[i]
+		if b < 0x20 || b > 0x7E {
+			return false
+		}
+		if b == ',' {
+			return false
```
[the remark this answers is not in the document]
Comment (rarguelloF) [reply internal/llmobs/util.go]: +1, confirmed with `DD_TRACE_PROPAGATION_STYLE=tracecontext`. Both `agent;v2` and `agent~v2` arrive downstream as `agent_v2`, so the reported attribution is a name that never existed. `agent=v2` round-trips fine because `composeTracestate` maps `=` to `~` and extraction reverses it, which is the one case the doc comment and `util_test.go` already cover. Rejecting `;` and `~` alongside the comma is the smallest fix.

File: internal/llmobs/llmobs_test.go [hunk 13]
```diff
@@ -2384,6 +2384,274 @@ func traceHandler(h http.Handler) http.Handler {
...
+		tracer.Flush()
+
+		attr, ok := coll.RequireSpan(t, "tool").Meta["agent_attribution"].(map[string]any)
+		require.True(t, ok, "tool with propagated agent ancestor must have agent_attribution")
+		assert.Equal(t, "remote_agent", attr["pagent_name"])
+		assert.Equal(t, "remote-agent-id", attr["pagent_span_id"])
+	})
+}
+
+func TestAgentAttributionCrossProcess(t *testing.T) {
```
Comment (rarguelloF) [inline internal/llmobs/llmobs_test.go]: `TestStartSpan` already has a `distributed-context-propagation-*` family, one subtest per propagated field, and these cases belong there rather than in a separate block that rebuilds the server setup. `testClientServer` at line 2359 gives you `traceHandler` plus a `tracedRT` client that starts the client span and injects, so each case drops to starting the LLM spans and calling `cl.Do`, and it exercises the same inject path a real contrib would instead of a manual `tracer.Inject`.

`distributed-context-propagation-clears-stale-session` is the closest model, since the stale attribution Codex flagged is the same failure for a different tag. Cases I'd want covered:

- a sibling with no agent ancestor after an agent published on the same trace, asserting no attribution downstream
- a tracecontext-only hop with `;` in the name
- the wire keys themselves, since cross-language compatibility rests on `_dd.p.llmobs_pagent_span_id` and `_dd.p.llmobs_pagent_name` matching py and js
- the real id-only path, where an unsafe name is dropped on the wire and the downstream span emits `pagent_name: null`. The `id-only-emits-null-name` case builds the `PropagatedLLMSpan` by hand instead
- a three service hop, confirming the middle service forwards the original agent rather than re-attributing

Worth noting `no-agent-upstream-omits-attribution` also discards its workflow span without finishing it, so that span never gets flushed.

File: internal/llmobs/llmobs_test.go [hunk 14]
```diff
@@ -2384,6 +2384,274 @@ func traceHandler(h http.Handler) http.Handler {
...
+		assert.Equal(t, "outer_agent", innerAttr["pagent_name"])
+		assert.Equal(t, outer.SpanID(), innerAttr["pagent_span_id"])
+
+		toolAttr, ok := coll.RequireSpan(t, "inner_tool").Meta["agent_attribution"].(map[string]any)
+		require.True(t, ok, "inner tool must attribute to inner agent (nearest)")
+		assert.Equal(t, "inner_agent", toolAttr["pagent_name"])
+		assert.Equal(t, inner.SpanID(), toolAttr["pagent_span_id"])
+	})
+
+	t.Run("propagated-parent-takes-precedence-over-local-when-no-local-agent", func(t *testing.T) {
```
Comment (rarguelloF) [inline internal/llmobs/llmobs_test.go]: nit: the name says the propagated parent takes precedence over the local one, but there's no local parent here, so nothing is competing. Something like `propagated-agent-is-inherited-when-no-local-parent` would describe it better.

File: internal/llmobs/util_test.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+// Copyright 2025 Datadog, Inc.
+
+package llmobs
+
+import (
+	"strings"
+	"testing"
+)
+
+func TestAgentNameWireSafe(t *testing.T) {
```
Comment (rarguelloF) [inline internal/llmobs/util_test.go]: nit: the rest of the package uses testify, so `assert.True(t, AgentNameWireSafe(...))` would fit the surrounding style better than the bare `t.Error` calls.

**Review by rarguelloF — COMMENTED** [review]
A few problems in the propagation wiring, and Codex caught the main ones. I've replied on those threads with reproductions from this branch. Separately, the cross-process tests don't use the `testClientServer` helper the other distributed tests here use, and a few cases are missing.

Please also update the PR description. It still documents the removed `InjectContext`/`ExtractContext` API and says there are no changes to `ddtrace/tracer/`, which isn't the case anymore.

File: internal/llmobs/util.go (hunk 3 above)
Comment (rarguelloF) [reply internal/llmobs/util.go]: Reopening, `util.go` is back on the branch. Now that the helper is called from `ddtrace/tracer` I'd move it next to `PropagatedLLMSpan` in `internal/llmobs/context.go`, which is the other thing the tracer reaches into for this feature. Worth asking whether it needs to exist at all, since `isValidPropagatableTag` already rejects the same characters and skips just the offending tag, so if it stays the length half should become the budget check from Codex's thread.

File: ddtrace/tracer/span.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: Fixed! Now explicitly unsetting both pagent keys when pagentID is empty, same pattern as session_id. Added a `distributed-context-propagation-clears-stale-agent-attribution` test to cover it.

[the remark this answers is not in the document]
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: Fixed! Added `propagatingTagsByteLen()` which sums `len(k)+1+len(v)` over the current propagating tags (same measurement `marshalPropagatingTags` uses internally), then checks against `DefaultMaxTagsHeaderLen` before setting the name tag. Also guarding the empty-name case now so `isValidPropagatableTag` can't stamp a propagation error on the root span.

File: internal/llmobs/util.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply internal/llmobs/util.go]: Fixed! Added `;` and `~` to the reject list alongside comma. The function also moved to `context.go` in this round.

File: internal/llmobs/llmobs_test.go (hunk 13 above)
Comment (yahya-mouman) [reply internal/llmobs/llmobs_test.go]: Done! Moved everything into `TestStartSpan/distributed-context-propagation-*` using `testClientServer`, and added the missing cases: stale attribution (`clears-stale-agent-attribution`), wire keys (`agent-attribution-wire-keys`), and id-only when name is unsafe (`id-only-when-name-unsafe`).

File: internal/llmobs/llmobs_test.go (hunk 14 above)
Comment (yahya-mouman) [reply internal/llmobs/llmobs_test.go]: Renamed to `propagated-agent-is-inherited-when-no-local-parent`

File: internal/llmobs/util_test.go (hunk 15 above)
Comment (yahya-mouman) [reply internal/llmobs/util_test.go]: `util_test.go` is gone, tests moved to `context_test.go` with `assert.True`/`assert.False` throughout.

**Comment by rarguelloF** [comment]
@codex review

File: ddtrace/tracer/span.go [hunk 16]
```diff
@@ -1347,6 +1348,29 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
...
+	// name whenever it is absent or wire-unsafe so isValidPropagatableTag can't stamp a propagation
+	// error for an empty value, and so a stale name from a previous sibling can't linger.
+	if pagentID := llmSpan.PropagatedParentAgentSpanID(); pagentID != "" {
+		spanCtx.trace.setPropagatingTag(keyPropagatedLLMObsPAgentSpanID, pagentID)
+		name := llmSpan.PropagatedParentAgentName()
+		if name != "" && illmobs.AgentNameWireSafe(name) {
+			// Only propagate the name if it fits in the remaining x-datadog-tags budget.
+			// This mirrors the len(k)+len(v) measurement used by marshalPropagatingTags.
+			used := spanCtx.trace.propagatingTagsByteLen()
+			if used+len(keyPropagatedLLMObsPAgentName)+len(name) <= internalconfig.DefaultMaxTagsHeaderLen {
```
[the remark this answers is not in the document]
Comment (darccio) [reply ddtrace/tracer/span.go]: Echoing on this comment and adding a suggestion:

The budget precheck uses the hardcoded default (`internalconfig.DefaultMaxTagsHeaderLen`)
instead of the actually configured limit. `marshalPropagatingTags` enforces
`p.cfg.MaxTagsHeaderLen`, which is configurable via `DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH`.
When that value is set lower than 512, this precheck under-guards, and on overflow
`marshalPropagatingTags` does `sb.Reset()` — dropping the *entire* x-datadog-tags header
(trace_id, parent_id, ml_app, sampling decision, everything), not just the agent name.

Separately (folding in a minor nit at the same spot): the byte count omits the `,` and `=`
separators that adding the tag actually costs (`len(k)+1+len(v)`, matching
`propagatingTagsByteLen`'s own measurement one line above) — off by 2 bytes, harmless in
practice but worth matching exactly since it's guarding the same budget.

```suggestion
			if used+len(keyPropagatedLLMObsPAgentName)+len(name)+2 <= internalconfig.Get().MaxTagsHeaderLen() {
```

File: ddtrace/tracer/span.go [hunk 17]
```diff
@@ -1347,6 +1348,29 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
 	} else {
 		spanCtx.trace.unsetPropagatingTag(keyPropagatedLLMObsSessionID)
 	}
+	// pagent tags are optional and trace-scoped. Unset both when there is no agent ancestor so a
+	// predecessor's values don't leak to downstream services. When there is an ancestor, unset the
+	// name whenever it is absent or wire-unsafe so isValidPropagatableTag can't stamp a propagation
+	// error for an empty value, and so a stale name from a previous sibling can't linger.
+	if pagentID := llmSpan.PropagatedParentAgentSpanID(); pagentID != "" {
+		spanCtx.trace.setPropagatingTag(keyPropagatedLLMObsPAgentSpanID, pagentID)
```
Comment (darccio) [inline ddtrace/tracer/span.go]: `keyPropagatedLLMObsPAgentSpanID` is set unconditionally here, with no budget check, while
only the name (added a few lines below) is gated on `propagatingTagsByteLen()`. If the goal
of the budget check is "never overflow the x-datadog-tags header," this undercuts it — the
always-on span_id tag can itself push an already-near-full header over the limit before the
name check even runs. If the intent is instead "span_id is essential, name is best-effort,"
that's a reasonable product call, but it's worth stating explicitly in the comment above
(lines 1351-1354) so a future reader doesn't assume both tags are budget-gated.

File: ddtrace/tracer/span.go [hunk 18]
```diff
@@ -1347,6 +1348,29 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
 	} else {
 		spanCtx.trace.unsetPropagatingTag(keyPropagatedLLMObsSessionID)
 	}
+	// pagent tags are optional and trace-scoped. Unset both when there is no agent ancestor so a
+	// predecessor's values don't leak to downstream services. When there is an ancestor, unset the
+	// name whenever it is absent or wire-unsafe so isValidPropagatableTag can't stamp a propagation
+	// error for an empty value, and so a stale name from a previous sibling can't linger.
```
Comment (darccio) [inline ddtrace/tracer/span.go]: Worth acknowledging a race this doesn't (and can't) close: propagating tags are trace-scoped
and shared, so two sibling spans under the same trace making *simultaneous* outbound calls
could pick up each other's pagent tags — the same property `session_id` already has. The
stale-clearing logic here correctly handles the *sequential* sibling case (and is well tested
for it), but a concurrent one isn't preventable with this data model. A short comment would
save the next reader from re-deriving this.

```suggestion
	// pagent tags are optional and trace-scoped. Unset both when there is no agent ancestor so a
	// predecessor's values don't leak to downstream services. When there is an ancestor, unset the
	// name whenever it is absent or wire-unsafe so isValidPropagatableTag can't stamp a propagation
	// error for an empty value, and so a stale name from a previous sibling can't linger.
	// Concurrent siblings under the same trace making outbound calls at the same time can still
	// race on these shared, trace-scoped tags (same caveat as session_id above) — this only
	// closes the sequential case.
```

**Review by darccio — CHANGES_REQUESTED** [review]
Two branches are untested:

- The over-budget name-drop `else` at ddtrace/tracer/span.go:1364-1366 — no test drives
  `propagatingTagsByteLen()` close enough to `MaxTagsHeaderLen` to exercise the drop path.
- The W3C tracestate `=` round-trip for a propagated agent name. `AgentNameWireSafe` explicitly
  allows `=` (internal/llmobs/context_test.go:25-27, "equals-sign-is-safe"), and
  `composeTracestate`/`sanitizeTagValue` encode it as `~` and decode it back — but no
  distributed test in internal/llmobs/llmobs_test.go actually sends an agent name containing
  `=` through a real tracestate hop to confirm it survives. The comma/unsafe-name id-only path
  (`distributed-context-propagation-id-only-when-name-unsafe`) is well covered; this is the
  complementary "safe special character actually round-trips" case.

Neither blocks merging given the surrounding test density, but both are cheap to add and
would close out the edge cases this PR itself calls out as deliberate (the `=` allowance,
the budget guard).

**Comment by yahya-mouman** [comment]
@codex

**Review by darccio — APPROVED** [review]
Thanks for addressing my feedback.

**Review by rarguelloF — APPROVED** [review]

File: ddtrace/tracer/span.go [hunk 19]
```diff
@@ -1347,6 +1348,61 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
...
+	// Concurrent siblings under the same trace making outbound calls at the same time can still
+	// race on these shared, trace-scoped tags (same caveat as session_id above) — this only
+	// closes the sequential case.
+	if pagentID := llmSpan.PropagatedParentAgentSpanID(); pagentID != "" {
+		// span_id is always set without a budget check: it is the essential half of
+		// attribution (id-only is still useful), and at ~43 bytes it is unlikely to
+		// overflow the header on its own. The name below is the best-effort half and
+		// is dropped when either budget is tight.
+		spanCtx.trace.setPropagatingTag(keyPropagatedLLMObsPAgentSpanID, pagentID)
+		name := llmSpan.PropagatedParentAgentName()
```
Comment (ncybul) [inline ddtrace/tracer/span.go]: Can we clean up this section of code? Possibly by moving this logic into a helper method and removing all the verbose comments which may help with readability.

It might also be a good idea to include a log or metric for whenever we drop the name (let's mirror whatever telemetry we have in python)

File: internal/llmobs/context.go [hunk 20]
```diff
@@ -50,3 +56,25 @@ func ActiveLLMSpanFromContext(ctx context.Context) (*Span, bool) {
...
+// AgentNameWireSafe reports whether name can safely be written as a
+// propagating-tag value. The rules match the shared cross-language contract:
+//
+//   - reject any byte outside the printable ASCII range [0x20, 0x7E]
+//   - reject comma (tagset delimiter) and semicolon / tilde (W3C tracestate
+//     characters sanitized by composeTracestate, which would corrupt attribution)
+//
+// The length check is omitted here; callers are responsible for verifying that
+// adding the name tag does not exceed the x-datadog-tags budget.
+func AgentNameWireSafe(name string) bool {
```
Comment (ncybul) [inline internal/llmobs/context.go:71]: Does this match the corresponding sanitization done in Node?

**Review by ncybul — APPROVED** [review]

File: ddtrace/tracer/span.go [hunk 21]
```diff
@@ -1347,6 +1348,50 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
...
 		spanCtx.trace.unsetPropagatingTag(keyPropagatedLLMObsSessionID)
 	}
+	setPropagatingPAgentTags(spanCtx, llmSpan.PropagatedParentAgentSpanID(), llmSpan.PropagatedParentAgentName())
+}
+
+// setPropagatingPAgentTags writes the pagent attribution tags onto the trace's propagating tag set.
+// Both tags are trace-scoped, so they are always explicitly set or unset to prevent stale values
+// from a previous sibling from leaking to downstream services (sequential case only — concurrent
+// siblings can still race, same caveat as session_id).
+func setPropagatingPAgentTags(spanCtx *SpanContext, pagentID, name string) {
```
Comment (mtoffl01) [inline ddtrace/tracer/span.go]: This, and the variables, looks like a typo to my eyes, whether or not it is. can it be renamed

File: ddtrace/tracer/span.go [hunk 22]
```diff
@@ -1347,6 +1348,50 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
...
+	if pagentID == "" {
+		spanCtx.trace.unsetPropagatingTag(keyPropagatedLLMObsPAgentSpanID)
```
Comment (mtoffl01) [inline ddtrace/tracer/span.go]: why do you need to unset it here if you are presumably overwriting the value unconditionally on line 1365?

Similar question for keyPropagatedLLMObsPAgentName, although that's not unconditional

File: ddtrace/tracer/span.go [hunk 23]
```diff
@@ -1347,6 +1348,50 @@ func setLLMObsPropagatingTags(ctx context.Context, spanCtx *SpanContext) {
...
+		return
+	}
+	// span_id is always propagated: id-only attribution is still useful to the backend.
+	spanCtx.trace.setPropagatingTag(keyPropagatedLLMObsPAgentSpanID, pagentID)
+	if name == "" || !illmobs.AgentNameWireSafe(name) {
+		spanCtx.trace.unsetPropagatingTag(keyPropagatedLLMObsPAgentName)
+		return
+	}
+	// Only propagate the name when it fits within both the x-datadog-tags and W3C tracestate budgets.
+	const tracestateHeaderReserve = 28 // fixed "dd=s:X;p:<hex>" prefix written by composeTracestate
```
Comment (mtoffl01) [inline ddtrace/tracer/span.go]: why defining a const inside of this function that is called numerous times? can it be defined once globally instead?

File: ddtrace/tracer/propagating_tags.go [hunk 24]
```diff
@@ -201,6 +202,41 @@ func (t *trace) propagatingTagsLen() int {
 	return len(t.loadPropagatingTags())
 }
 
+// propagatingTagsByteLen returns the byte length of the serialized propagating-tags
+// header value (format: "k1=v1,k2=v2,..."), matching the measurement used by
+// marshalPropagatingTags so callers can pre-check whether a new tag fits in the budget.
+// Like marshalPropagatingTags, it excludes tracestate/traceparent — those are stored in
+// the same map for W3C propagation but are never written to the x-datadog-tags header.
+func (t *trace) propagatingTagsByteLen() int {
```
Comment (mtoffl01) [inline ddtrace/tracer/propagating_tags.go]: So we have to recalculate this on every span? An alternative is to maintain another variable that tracks the length of the propagating tags; or change the type of the propagating tags to a custom struct that reports its own length

File: ddtrace/tracer/span.go (hunk 21 above)
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: I can rename the var however the tag would remain ``pagent_name``, ``pagent_span_id``. As that's what we have on JS and Python SDKs (initially to reduce tag footprint).

File: ddtrace/tracer/propagating_tags.go (hunk 24 above)
Comment (yahya-mouman) [reply ddtrace/tracer/propagating_tags.go]: the map has fewer than 10 entries in practice (bounded by spec), so iterating it is ~10 string-length lookups per outbound span. threading a cached length through every `setPropagatingTag`/`unsetPropagatingTag` callsite felt like more complexity than the savings justify. added a note to the doc comment to make this explicit.

File: internal/llmobs/context.go (hunk 20 above)
Comment (yahya-mouman) [reply internal/llmobs/context.go:71]: the divergence from node is intentional. node only routes through x-datadog-tags, so `;'` and `~` are fine there. in go, names also travel through W3C tracestate where `composeTracestate` sanitizes both to `_`, meaning `agent;v2` would arrive downstream as `agent_v2`. rejecting those chars upfront seemed cleaner than letting the wire shape what attribution value the backend sees.

File: ddtrace/tracer/span.go (hunk 21 above)
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: renamed `setPropagatingPAgentTags` → `setPropagatingParentAgentTags`, parameter `pagentID` → `parentAgentSpanID`, and the key constants `keyPropagatedLLMObsPAgentSpanID`/`Name` → `keyPropagatedLLMObsParentAgentSpanID`/`Name`. wire values stay the same since they're protocol-defined.

File: ddtrace/tracer/span.go (hunk 22 above)
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: you're right that it reads that way. restructured to `if/else` so the two branches are visually exclusive.

File: ddtrace/tracer/span.go (hunk 23 above)
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: moved to the `textmap.go` constants block alongside `tracestateDDMaxSize`.

File: ddtrace/tracer/span.go (hunk 19 above)
Comment (yahya-mouman) [reply ddtrace/tracer/span.go]: extracted into `setPropagatingParentAgentTags` and verbose comments trimmed. the `log.Debug` on name drop matches what python does (debug-level only, no metric).

File: ddtrace/tracer/propagating_tags.go (hunk 24 above)
Comment (mtoffl01) [reply ddtrace/tracer/propagating_tags.go]: What about combining the logic in `propagatingTagsByteLen()` and `propagatingTagsTracestateByteLen()` into a single call to avoid duplicate passes over the same map on every call?

**Review by mtoffl01 — APPROVED** [review]
Approving with one request for an optimization at the propagatingTagsByteLen layer
